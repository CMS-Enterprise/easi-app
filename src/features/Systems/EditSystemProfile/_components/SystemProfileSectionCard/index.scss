.system-profile-section-card {
  &__description {
    display: -webkit-box;
    -webkit-line-clamp: 4;
    line-clamp: 4;
    -webkit-box-orient: vertical;
    overflow: hidden;
  }
}
