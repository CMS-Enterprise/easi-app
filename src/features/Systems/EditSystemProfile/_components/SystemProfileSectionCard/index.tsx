import React from 'react';
import { useTranslation } from 'react-i18next';
import { Link, useParams } from 'react-router-dom';
import {
  Card,
  CardBody,
  CardFooter,
  CardHeader,
  Icon
} from '@trussworks/react-uswds';
import classNames from 'classnames';
import { SystemProfileSectionLockStatus } from 'gql/generated/graphql';
import { useFlags } from 'launchdarkly-react-client-sdk';

import PercentCompleteTag from 'components/PercentCompleteTag';
import SectionLock from 'components/SectionLock';
import { SystemProfileSection } from 'types/systemProfile';

import { getSystemProfileSectionMap } from '../../util';
import ExternalDataTag from '../ExternalDataTag';

import './index.scss';

type SystemProfileSectionCardProps = {
  section: SystemProfileSection;
  percentComplete?: number;
  hasPendingChanges?: boolean;
  isManagedExternally?: boolean;
  hasExternalData?: boolean;
  readOnly?: boolean;
};

/**
 * Displays information about a system profile section with view/edit button
 */
const SystemProfileSectionCard = ({
  section,
  percentComplete,
  hasPendingChanges,
  isManagedExternally,
  hasExternalData,
  readOnly
}: SystemProfileSectionCardProps) => {
  const { t } = useTranslation('systemProfile');

  const { systemId } = useParams<{
    systemId: string;
  }>();

  const flags = useFlags();
  const sectionMap = getSystemProfileSectionMap(flags);
  const { route } = sectionMap[section];

  // TODO EASI-4984: Update to use actual section lock context
  const lockableSectionLocks = [] as SystemProfileSectionLockStatus[];

  /** Returns lock status if section is locked */
  const sectionLock: SystemProfileSectionLockStatus | undefined =
    lockableSectionLocks?.find(lock => lock.section === section);

  return (
    <Card
      className="tablet:grid-col-6 desktop:grid-col-4"
      containerProps={{ className: 'shadow-2 border-width-1px' }}
      data-testid={`section-card-${section}`}
    >
      {isManagedExternally && (
        <h5 className="text-base-dark bg-base-lightest padding-x-3 margin-0 font-body-xs text-light display-flex flex-align-center">
          <Icon.InfoOutline className="margin-right-05" aria-hidden />
          {t(
            readOnly
              ? 'editSystemProfile.dataManagedExternally'
              : 'editSystemProfile.dataPartiallyManagedExternally'
          )}
        </h5>
      )}

      <CardHeader
        className={classNames(
          'padding-bottom-0',
          isManagedExternally && 'padding-top-105'
        )}
      >
        <h3>{t(`sectionCards.${section}.title`)}</h3>
      </CardHeader>

      <CardBody
        className={classNames(
          'padding-top-05',
          !hasPendingChanges && 'padding-bottom-2'
        )}
      >
        <p className="system-profile-section-card__description">
          {t(`sectionCards.${section}.description`)}
        </p>

        {percentComplete !== undefined && (
          <PercentCompleteTag percentComplete={percentComplete} />
        )}

        {hasExternalData !== undefined && (
          <ExternalDataTag hasExternalData={hasExternalData} />
        )}
      </CardBody>

      <CardFooter>
        {hasPendingChanges && (
          <p className="text-base-dark margin-top-1 margin-bottom-0 display-flex flex-align-center">
            <Icon.Schedule className="margin-right-05" aria-hidden />
            {t('editSystemProfile.sectionHasPendingChanges')}
          </p>
        )}

        {sectionLock ? (
          <SectionLock sectionLock={sectionLock} />
        ) : (
          <Link
            to={`/systems/${systemId}/${route}`}
            className={`usa-button ${
              hasPendingChanges ? 'usa-button--unstyled' : 'usa-button--outline'
            }`}
          >
            {hasPendingChanges || readOnly
              ? t('editSystemProfile.viewSection')
              : t('editSystemProfile.editSection')}
          </Link>
        )}
      </CardFooter>
    </Card>
  );
};

export default SystemProfileSectionCard;
