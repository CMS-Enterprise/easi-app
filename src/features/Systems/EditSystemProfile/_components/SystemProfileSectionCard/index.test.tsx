import React from 'react';
import { render, screen } from '@testing-library/react';

import SystemProfileSectionCard from '.';

describe('SystemProfileSectionCard', () => {
  it('matches the snapshot', () => {
    const { asFragment } = render(
      <SystemProfileSectionCard
        title="Section Title"
        description="This is the section card description."
        route="section-route"
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders partially managed externally', () => {
    render(
      <SystemProfileSectionCard
        title="Section Title"
        description="This is the section card description."
        route="section-route"
        isManagedExternally
      />
    );

    expect(
      screen.getByText('Data partially managed externally')
    ).toBeInTheDocument();

    expect(screen.getByRole('button', { name: 'Edit section' })).toHaveClass(
      'usa-button--outline'
    );
  });

  it('renders managed externally - read only view', () => {
    render(
      <SystemProfileSectionCard
        title="Section Title"
        description="This is the section card description."
        route="section-route"
        isManagedExternally
        readOnly
      />
    );

    expect(screen.getByText('Data managed externally')).toBeInTheDocument();

    expect(screen.getByRole('button', { name: 'View section' })).toHaveClass(
      'usa-button--outline'
    );
  });

  it('renders section with pending changes ', () => {
    render(
      <SystemProfileSectionCard
        title="Section Title"
        description="This is the section card description."
        route="section-route"
        hasPendingChanges
      />
    );

    expect(
      screen.getByText('This section has pending changes')
    ).toBeInTheDocument();

    expect(screen.getByRole('button', { name: 'View section' })).toHaveClass(
      'usa-button--unstyled'
    );
  });
});
