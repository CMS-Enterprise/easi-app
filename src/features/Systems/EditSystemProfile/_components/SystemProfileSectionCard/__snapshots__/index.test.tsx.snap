// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SystemProfileSectionCard > matches the snapshot 1`] = `
<DocumentFragment>
  <li
    class="usa-card grid-col-4"
    data-testid="Card"
  >
    <div
      class="usa-card__container"
    >
      <div
        class="usa-card__header padding-bottom-0"
        data-testid="CardHeader"
      >
        <h3>
          Section Title
        </h3>
      </div>
      <div
        class="usa-card__body padding-top-05"
        data-testid="CardBody"
      >
        <p>
          This is the section card description.
        </p>
      </div>
      <div
        class="usa-card__footer"
        data-testid="CardFooter"
      >
        <button
          class="usa-button usa-button--outline"
          data-testid="button"
          type="button"
        >
          Edit section
        </button>
      </div>
    </div>
  </li>
</DocumentFragment>
`;
